sdks/python/apache_beam/io/gcp/bigtableio.py

```diff
@@ -141,3 +148,129 @@ def expand(self, pvalue):
+    			  calling ``cancel()``.
+
+    ***** NOTE: For unclear reasons, the function returns generator even
+    after wrapping the result as a list. In order to be used as a list, the
+    result should be wrapped as a list AGAIN! E.g., see 'estimate_size()'
+    """
+    if self.sample_row_keys is None:
+      self.sample_row_keys = list(self._get_table().sample_row_keys())
+    return self.sample_row_keys
+
+  def get_range_tracker(self, start_position=b'', stop_position=b''):
+    if stop_position == b'':
+      return LexicographicKeyRangeTracker(start_position)
+    else:
+      return LexicographicKeyRangeTracker(start_position, stop_position)
+
+  def estimate_size(self):
+    return list(self.get_sample_row_keys())[-1].offset_bytes
+
+  def split(self, desired_bundle_size, start_position=None, stop_position=None):
+    """ Splits the source into a set of bundles, using the row_set if it is set.
+
+    *** At this point, only splitting an entire table into samples based on the sample row keys is supported ***
+
+    :param desired_bundle_size: the desired size (in bytes) of the bundles returned.
+    :param start_position: if specified, the position must be used as the starting position of the first bundle.
+    :param stop_position: if specified, the position must be used as the ending position of the last bundle.
+    Returns:
+    	an iterator of objects of type 'SourceBundle' that gives information about the generated bundles.
+    """
+
+    if start_position is not None or stop_position is not None:
+      raise NotImplementedError
+
+    # TODO: Use the desired bundle size to split accordingly
+    # TODO: Allow users to provide their own row sets
+
+    sample_row_keys = list(self.get_sample_row_keys())
+
+    if len(sample_row_keys) > 1 and sample_row_keys[0].row_key != b'':
+      SampleRowKey = namedtuple("SampleRowKey", "row_key offset_bytes")
+      first_key = SampleRowKey(b'', 0)
+      sample_row_keys.insert(0, first_key)
+      sample_row_keys = list(sample_row_keys)
+
+    bundles = []
+    for i in range(1, len(sample_row_keys)):
+      key_1 = sample_row_keys[i - 1].row_key
+      key_2 = sample_row_keys[i].row_key
+      size = sample_row_keys[i].offset_bytes - sample_row_keys[i - 1].offset_bytes
+      bundles.append(iobase.SourceBundle(size, self, key_1, key_2))
+
+    # Shuffle is needed to allow reading from different locations of the table for better efficiency
+    shuffle(bundles)
+    return bundles
+
+  def read(self, range_tracker):
+    for row in self._get_table().read_rows(start_key=range_tracker.start_position(),
+                                           end_key=range_tracker.stop_position(),
+                                           filter_=self.beam_options['filter_']):
+      if range_tracker.try_claim(row.row_key):
+        self.row_count.inc()
+        yield row
+      else:
+        # TODO: Modify the client ot be able to cancel read_row request
```

Comment:
s/ot/to

Also please create a JIRA and mention in the TODO.

Reply:
Although not applicable to the current scheme, the Bigtable client has already been updated, see issue [7760](https://github.com/googleapis/google-cloud-python/issues/7760).